### ISSUE TITLE
TokenGS: Decoupling 3D Gaussian Prediction from Pixels with Learnable Tokens

### ISSUE THREAD

**Issue by WoojunePark** [WoojunePark/splat_papers#44]
## [TokenGS: Decoupling 3D Gaussian Prediction from Pixels with Learnable Tokens](https://arxiv.org/abs/2604.15239v1)

**Status:** `to-read` | **Published:** 2026-04-16

### Abstract

> In this work, we revisit several key design choices of modern Transformer-based approaches for feed-forward 3D Gaussian Splatting (3DGS) prediction. We argue that the common practice of regressing Gaussian means as depths along camera rays is suboptimal, and instead propose to directly regress 3D mean coordinates using only a self-supervised rendering loss. This formulation allows us to move from the standard encoder-only design to an encoder-decoder architecture with learnable Gaussian tokens, thereby unbinding the number of predicted primitives from input image resolution and number of views. Our resulting method, TokenGS, demonstrates improved robustness to pose noise and multiview inconsistencies, while naturally supporting efficient test-time optimization in token space without degrading learned priors. TokenGS achieves state-of-the-art feed-forward reconstruction performance on both static and dynamic scenes, producing more regularized geometry and more balanced 3DGS distribution, while seamlessly recovering emergent scene attributes such as static-dynamic decomposition and scene flow.

### Links

[arXiv](https://arxiv.org/abs/2604.15239v1) | [PDF](https://arxiv.org/pdf/2604.15239v1) | [HTML](https://arxiv.org/html/2604.15239v1) | [Website](https://research.nvidia.com/labs/toronto-ai/tokengs) | [Code](https://github.com/nv-tlabs/TokenGS)

### Figures

![Figure](https://arxiv.org/html/2604.15239v1/x1.png)
*Figure 1: TokenGS is a feed-forward reconstruction framework that outputs a 3D Gaussian Splatting (3DGS) representation from posed input images. Our novel encoder-decoder architecture detaches 3D Gaussians from input pixels and enables multiple properties desirable for 3D reconstruction, as demonstrated in each example.*

![Figure](https://arxiv.org/html/2604.15239v1/x2.png)
*Figure 2: Our method. We design a novel network architecture that reconstructs 3DGS from input images, directly predicting 3D Gaussian mean coordinates. The model follows an encoder-decoder structure. In the decoder, 3DGS tokens are fed in as queries to obtain the final Gaussian attributes. After the base model is trained, we allow test-time token tuning from input images to improve reconstruction quality.*

![Figure](https://arxiv.org/html/2604.15239v1/x3.png)
*Figure 3: Comparison of 3DGS parametrizations in feed-forward networks. While all methods reconstruct the single input view well (shown in the inset), the quality of the occluded region behind the table varies.*

![Figure](https://arxiv.org/html/2604.15239v1/x4.png)
*Figure*

![Figure](https://arxiv.org/html/2604.15239v1/x5.png)
*Figure 4: Attention masking for dynamic scenes. The figure shows one instantiation of the 3DGS self-attention block. The horizontal axis shows queries, while the vertical axis shows keys/values of the corresponding tokens.*

![Figure](https://arxiv.org/html/2604.15239v1/x6.png)
*Figure 5: Qualitative results on RE10K [56]. Compared to the pixel-aligned Gaussian prediction of GS-LRM [54], our formulation, which is based on direct XYZ prediction and decoupling from pixel rays, produces noticeably cleaner geometry with fewer spiky artifacts.*

![Figure](https://arxiv.org/html/2604.15239v1/x7.png)
*Figure 6: View Extrapolation. Both GS-LRM and our model have been finetuned for view extrapolation. The lower-left inset figures show the ground-truth images.*

![Figure](https://arxiv.org/html/2604.15239v1/x8.png)
*Figure 7: Reconstruction under camera noise on 2-view RE10K. We add camera pose noise of magnitude 1–10 degrees to the non-reference view. We show the difference to GS-LRM [54] in terms of PSNR and LPIPS. Note, that we visualize −Δ​LPIPS-\Delta\text{LPIPS} so higher values indicate better performance for both metrics.*

![Figure](https://arxiv.org/html/2604.15239v1/x9.png)
*Figure 8: Dynamic reconstruction on Kubric [13]. We show in-between time stamps to examine temporal consistency.*

![Figure](https://arxiv.org/html/2604.15239v1/x10.png)
*Figure 9: Emergent scene flow. We visualize the trajectories of each dynamic Gaussian across time.*

![Figure](https://arxiv.org/html/2604.15239v1/figures/gaussian_token_binding/regular_rendering.png)
*Figure 10: Token assignments. Left: rendering. Center: GS from 5 random tokens highlighted with per-token colors to show their location across scenes. Others set to gray; overlaid with GT image. Right: GS from each token assigned a consistent color.*

![Figure](https://arxiv.org/html/2604.15239v1/x11.png)
*Figure 11: Effect of the visibility loss. Regularizing the model removes floaters from unobserved parts of the scene and improves the final PSNR on novel views.*

![Figure](https://arxiv.org/html/2604.15239v1/x12.png)
*Figure 12: Test-time scaling: PSNR when scaling #input views vs #gradient steps in two variants: optimizing Gaussian tokens (solid) or Gaussian parameters directly (dashed).*

![Figure](https://arxiv.org/html/2604.15239v1/x13.png)
*Figure 13: Center: renders of a feed-forward reconstruction from 4 views, completely novel viewpoint. Left: Gaussian Tuning degrades scene geometry, despite quantitative advantage on close-by views. Right: Our Token Tuning (TT) improves scene geometry with sharper renderings.*

![Figure](https://arxiv.org/html/2604.15239v1/x14.png)
*Figure 14: PSNR vs # Gaussian tokens. We study how reconstruction quality changes with different numbers of Gaussian tokens given fixed number of input views.*

---

### 📝 Reading Notes Template

Leave a comment using the sections below.
Bullet items under each heading are synced to the paper's YAML tags.
Any other text goes to `## My Notes`.
Heading level (`#` or `##`) and singular/plural spelling both work.

```
## inputs
- 

## outputs
- 

## methods
- 

Free-form reading notes here.
```

*Managed by splat-papers. Close this issue once you have read the paper.*
